<a href="https://colab.research.google.com/github/satyamkarn100-ctrl/Neural-Recommendation-Personalization-Engine/blob/main/02_preprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
!pip install -q huggingface_hub
from huggingface_hub import hf_hub_download
from huggingface_hub import login
login()

In [2]:
reviews_path = hf_hub_download(
    repo_id="Satyamkarn100/NeuraRec-Electronics",
    filename="raw/reviews.parquet",
    repo_type="dataset"
)

raw/reviews.parquet: reconstructing file:   0%|          |  0.00B /  313MB            

raw/reviews.parquet: downloading bytes:           |  0.00B            

In [3]:
reviews = pd.read_parquet(reviews_path)
print('Reviews Shape:',reviews.shape)


Reviews Shape: (15473536, 4)


In [4]:
reviews.head()

,user_id,parent_asin,rating,timestamp
0,AGCI7FAH4GL5FI65HYLKWTMFZ2CQ,B0047T79VS,3.0,1344406083000
1,AGCI7FAH4GL5FI65HYLKWTMFZ2CQ,B01HHURN3W,3.0,1408995743000
2,AGCI7FAH4GL5FI65HYLKWTMFZ2CQ,B00L0YLRUW,1.0,1439226089000
3,AGCI7FAH4GL5FI65HYLKWTMFZ2CQ,B017T99JPG,5.0,1456772365000
4,AGCI7FAH4GL5FI65HYLKWTMFZ2CQ,B01LW71IBJ,5.0,1456772571000


In [5]:
reviews['datetime'] = pd.to_datetime(
    reviews['timestamp'],unit='ms'
)

In [6]:
print(reviews['parent_asin'].nunique())
print()
reviews.head(10)

368228



,user_id,parent_asin,rating,timestamp,datetime
0,AGCI7FAH4GL5FI65HYLKWTMFZ2CQ,B0047T79VS,3.0,1344406083000,2012-08-08 06:08:03.000
1,AGCI7FAH4GL5FI65HYLKWTMFZ2CQ,B01HHURN3W,3.0,1408995743000,2014-08-25 19:42:23.000
2,AGCI7FAH4GL5FI65HYLKWTMFZ2CQ,B00L0YLRUW,1.0,1439226089000,2015-08-10 17:01:29.000
3,AGCI7FAH4GL5FI65HYLKWTMFZ2CQ,B017T99JPG,5.0,1456772365000,2016-02-29 18:59:25.000
4,AGCI7FAH4GL5FI65HYLKWTMFZ2CQ,B01LW71IBJ,5.0,1456772571000,2016-02-29 19:02:51.000
5,AGCI7FAH4GL5FI65HYLKWTMFZ2CQ,B09S6Y5BRG,5.0,1541356831659,2018-11-04 18:40:31.659
6,AGCI7FAH4GL5FI65HYLKWTMFZ2CQ,B07BHHB5RH,5.0,1565130879386,2019-08-06 22:34:39.386
7,AGCI7FAH4GL5FI65HYLKWTMFZ2CQ,B07CML419K,5.0,1637522881041,2021-11-21 19:28:01.041
8,AGBFYI2DDIKXC5Y4FARTYDTQBMFQ,B013KPXNSW,5.0,1453856042000,2016-01-27 00:54:02.000
9,AGBFYI2DDIKXC5Y4FARTYDTQBMFQ,B08P3VMW76,5.0,1483327495000,2017-01-02 03:24:55.000


In [7]:
meta_paths = []

for i in range(1, 11):
    path = hf_hub_download(
        repo_id="Satyamkarn100/NeuraRec-Electronics",
        filename=f"raw/meta{i}.parquet",
        repo_type="dataset"
    )
    meta_paths.append(path)

print("Downloaded:", len(meta_paths), "files")

raw/meta1.parquet: reconstructing file:   0%|          |  0.00B /  148MB            

raw/meta1.parquet: downloading bytes:           |  0.00B            

raw/meta2.parquet: reconstructing file:   0%|          |  0.00B /  145MB            

raw/meta2.parquet: downloading bytes:           |  0.00B            

raw/meta3.parquet: reconstructing file:   0%|          |  0.00B /  143MB            

raw/meta3.parquet: downloading bytes:           |  0.00B            

raw/meta4.parquet: reconstructing file:   0%|          |  0.00B /  142MB            

raw/meta4.parquet: downloading bytes:           |  0.00B            

raw/meta5.parquet: reconstructing file:   0%|          |  0.00B /  139MB            

raw/meta5.parquet: downloading bytes:           |  0.00B            

raw/meta6.parquet: reconstructing file:   0%|          |  0.00B /  134MB            

raw/meta6.parquet: downloading bytes:           |  0.00B            

raw/meta7.parquet: reconstructing file:   0%|          |  0.00B /  131MB            

raw/meta7.parquet: downloading bytes:           |  0.00B            

raw/meta8.parquet: reconstructing file:   0%|          |  0.00B /  134MB            

raw/meta8.parquet: downloading bytes:           |  0.00B            

raw/meta9.parquet: reconstructing file:   0%|          |  0.00B /  131MB            

raw/meta9.parquet: downloading bytes:           |  0.00B            

raw/meta10.parquet: reconstructing file:   0%|          |  0.00B /  126MB            

raw/meta10.parquet: downloading bytes:           |  0.00B            

Downloaded: 10 files


In [8]:
meta = pd.concat(
    [pd.read_parquet(path) for path in meta_paths],
    ignore_index=True
)

print("Meta-Data Shape:", meta.shape)
meta.head()

Meta-Data Shape: (1610012, 11)


,main_category,title,average_rating,rating_number,features,description,price,store,categories,details,parent_asin
0,All Electronics,FS-1051 FATSHARK TELEPORTER V3 HEADSET,3.5,6,[],[Teleporter V3 The “Teleporter V3” kit sets a ...,None,Fat Shark,"[Electronics, Television & Video, Video Glasses]","{""Date First Available"": ""August 2, 2014"", ""Ma...",B00MCW7G9M
1,All Electronics,Ce-H22B12-S1 4Kx2K Hdmi 4Port,5.0,1,"[UPC: 662774021904, Weight: 0.600 lbs]",[HDMI In - HDMI Out],None,SIIG,"[Electronics, Television & Video, Accessories,...","{""Product Dimensions"": ""0.83 x 4.17 x 2.05 inc...",B00YT6XQSE
2,Computers,Digi-Tatoo Decal Skin Compatible With MacBook ...,4.5,246,[WARNING: Please IDENTIFY MODEL NUMBER on the ...,[],19.99,Digi-Tatoo,"[Electronics, Computers & Accessories, Laptop ...","{""Brand"": ""Digi-Tatoo"", ""Color"": ""Fresh Marble...",B07SM135LS
3,AMAZON FASHION,NotoCity Compatible with Vivoactive 4 band 22m...,4.5,233,[☛NotoCity 22mm band is designed for Vivoactiv...,[],9.99,NotoCity,"[Electronics, Wearable Technology, Clips, Arm ...","{""Date First Available"": ""May 29, 2020"", ""Manu...",B089CNGZCW
4,Cell Phones & Accessories,Motorola Droid X Essentials Combo Pack,3.8,64,"[New Droid X Essentials Combo Pack, Exclusive ...",[all Genuine High Quality Motorola Made Access...,14.99,Verizon,"[Electronics, Computers & Accessories, Compute...","{""Product Dimensions"": ""11.6 x 6.9 x 3.1 inche...",B004E2Z88O


In [9]:
print("Metadata Shape:", meta.shape)
print("\nColumns:")
print(meta.columns.tolist())

print("\nData Types:")
print(meta.dtypes)

print("\nInfo:")
meta.info()

Metadata Shape: (1610012, 11)

Columns:
['main_category', 'title', 'average_rating', 'rating_number', 'features', 'description', 'price', 'store', 'categories', 'details', 'parent_asin']

Data Types:
main_category      object
title              object
average_rating    float64
rating_number       int64
features           object
description        object
price              object
store              object
categories         object
details            object
parent_asin        object
dtype: object

Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1610012 entries, 0 to 1610011
Data columns (total 11 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   main_category   1503678 non-null  object 
 1   title           1610012 non-null  object 
 2   average_rating  1610012 non-null  float64
 3   rating_number   1610012 non-null  int64  
 4   features        1610012 non-null  object 
 5   description     1610012 non-null  object 
 6  

In [10]:
missing = pd.DataFrame({
    "Missing_Count": meta.isnull().sum(),
    "Missing_Percentage": (meta.isnull().sum() / len(meta) * 100).round(2)
})

missing.sort_values("Missing_Count", ascending=False)

,Missing_Count,Missing_Percentage
main_category,106334,6.60
store,9520,0.59
title,0,0.00
average_rating,0,0.00
rating_number,0,0.00
description,0,0.00
features,0,0.00
price,0,0.00
categories,0,0.00
details,0,0.00


In [11]:
print("Price available:", meta["price"].notna().sum())
print("Price missing:", meta["price"].isna().sum())

Price available: 1610012
Price missing: 0


### Missing Value Observations

- The `price` column has 67.30% missing values, making it unreliable for direct use without careful handling.
- `main_category` has 6.60% missing values, indicating some products lack category information.
- `store` has only 0.59% missing values and is mostly complete.
- Core recommendation-related fields such as `parent_asin`, `average_rating`, `rating_number`, `title`, `description`, and `categories` have no missing values.
- Since price is missing for most products, it should not be used as a primary feature unless an appropriate missing-value strategy is applied.

In [14]:
meta['price'] = meta['price'].replace(
    ['None','none','NULL','null',''],np.nan
)

# Convert price to numeric
meta['price'] = pd.to_numeric(meta['price'], errors='coerce')

# Missing indicator
meta['price_missing'] = meta['price'].isna().astype(int)

before = meta['price'].isna().sum()
print("Missing before:", before)

# Same title based median imputation
title_median = meta.groupby('title')['price'].transform('median')

meta['price'] = meta['price'].fillna(title_median)

after = meta['price'].isna().sum()
print("Missing after title-match fill:", after)
print("Recovered rows:", before - after)

Missing before: 1083563
Missing after title-match fill: 1075867
Recovered rows: 7696


In [19]:
meta.head()

,main_category,title,average_rating,rating_number,features,description,price,store,categories,details,parent_asin,price_missing
0,All Electronics,FS-1051 FATSHARK TELEPORTER V3 HEADSET,3.5,6,[],[Teleporter V3 The “Teleporter V3” kit sets a ...,NaN,Fat Shark,"[Electronics, Television & Video, Video Glasses]","{""Date First Available"": ""August 2, 2014"", ""Ma...",B00MCW7G9M,1
1,All Electronics,Ce-H22B12-S1 4Kx2K Hdmi 4Port,5.0,1,"[UPC: 662774021904, Weight: 0.600 lbs]",[HDMI In - HDMI Out],NaN,SIIG,"[Electronics, Television & Video, Accessories,...","{""Product Dimensions"": ""0.83 x 4.17 x 2.05 inc...",B00YT6XQSE,1
2,Computers,Digi-Tatoo Decal Skin Compatible With MacBook ...,4.5,246,[WARNING: Please IDENTIFY MODEL NUMBER on the ...,[],19.99,Digi-Tatoo,"[Electronics, Computers & Accessories, Laptop ...","{""Brand"": ""Digi-Tatoo"", ""Color"": ""Fresh Marble...",B07SM135LS,0
3,AMAZON FASHION,NotoCity Compatible with Vivoactive 4 band 22m...,4.5,233,[☛NotoCity 22mm band is designed for Vivoactiv...,[],9.99,NotoCity,"[Electronics, Wearable Technology, Clips, Arm ...","{""Date First Available"": ""May 29, 2020"", ""Manu...",B089CNGZCW,0
4,Cell Phones & Accessories,Motorola Droid X Essentials Combo Pack,3.8,64,"[New Droid X Essentials Combo Pack, Exclusive ...",[all Genuine High Quality Motorola Made Access...,14.99,Verizon,"[Electronics, Computers & Accessories, Compute...","{""Product Dimensions"": ""11.6 x 6.9 x 3.1 inche...",B004E2Z88O,0


In [18]:
print(meta["main_category"].value_counts())

main_category
Computers                       418868
All Electronics                 376435
Camera & Photo                  223690
Cell Phones & Accessories       138237
Home Audio & Theater            106516
Industrial & Scientific          50555
Car Electronics                  27332
Tools & Home Improvement         24042
Amazon Home                      21299
Office Products                  21250
AMAZON FASHION                   18391
Sports & Outdoors                14581
Automotive                       13956
GPS & Navigation                  9339
Amazon Devices                    8484
Portable Audio & Accessories      8050
Musical Instruments               6668
Toys & Games                      3944
Health & Personal Care            3073
All Beauty                        1848
Arts, Crafts & Sewing             1496
Apple Products                    1078
Video Games                       1042
Books                              994
Baby                               797
Pet Supplie

In [23]:
computers = meta[
    meta['main_category'].eq('Computers')
]
computers.head(50)

,main_category,title,average_rating,rating_number,features,description,price,store,categories,details,parent_asin,price_missing
2,Computers,Digi-Tatoo Decal Skin Compatible With MacBook ...,4.5,246,[WARNING: Please IDENTIFY MODEL NUMBER on the ...,[],19.99,Digi-Tatoo,"[Electronics, Computers & Accessories, Laptop ...","{""Brand"": ""Digi-Tatoo"", ""Color"": ""Fresh Marble...",B07SM135LS,0
8,Computers,MOSISO Plastic Hard Shell Case & Keyboard Cove...,5.0,2,[],[],NaN,MOSISO,"[Electronics, Headphones, Earbuds & Accessorie...","{""Brand"": ""MOSISO"", ""Item Weight"": ""8 ounces"",...",B01MCZP7RF,1
10,Computers,Network Magic 4.0,3.3,14,"[Connect, Manage, and Secure Your Network, Sha...","[Product description, You try to connect your ...",NaN,Avanquest,"[Electronics, Computers & Accessories, Network...","{""Brand"": ""Avanquest"", ""Item model number"": ""4...",B000LV7P8I,1
13,Computers,eForCity Leather Case with Stand for 7-Inch Sa...,3.9,144,[Stop worrying about scratching your Samsung; ...,[Compatible With Samsung Galaxy Tab 2 P3100 7....,NaN,eForCity,"[Electronics, Computers & Accessories, Tablet ...","{""Standing screen display size"": ""7 Inches"", ""...",B00ABEQ6TY,1
14,Computers,Novelty Cute Cartoon USB 2.0 Flash Drive Data ...,4.7,198,[♥Best Gift♥-Catoon shape design different fro...,[This is a special usb stick that make as a cu...,8.98,Fire Stone,"[Electronics, Computers & Accessories, Data St...","{""Product Dimensions"": ""2.48 x 1.58 x 0.78 inc...",B07Y6K1PRB,0
17,Computers,BIPRA U3 2.5 inch USB 3.0 NTFS Portable Extern...,5.0,1,"[Storage capacity: 250GB, Formatted as NTFS fi...","[Bipra external USB 2.5"" portable size hard dr...",98.49,BIPRA,"[Electronics, Computers & Accessories, Data St...","{""Hard Drive"": ""250 GB usb 3"", ""Brand"": ""BIPRA...",B00VBTJCJE,0
18,Computers,WD Scorpio Blue 320GB Mobile SATA Hard Drive,4.0,37,[Low profile design to fit thin laptops and ta...,[WD Scorpio Blue 7 mm slim 2.5-inch SATA hard ...,NaN,Western Digital,"[Electronics, Computers & Accessories, Compute...","{""Hard Drive"": ""320 GB"", ""Brand"": ""Western Dig...",B007U1YGOS,1
20,Computers,"Gateway 15.6"" FHD Ultra Slim Budget Notebook, ...",4.1,15,"[【Processor】4 Core, 4 Threads, 4MB Cache, up t...","[Processor:, Intel® Pentium® Silver N5030 Proc...",189.99,Gateway,"[Electronics, Computers & Accessories, Compute...","{""Standing screen display size"": ""15.6 Inches""...",B0BYBG1PPD,0
26,Computers,HTC Vivid Black/Blue Perforated Cover,5.0,2,[Combines hard plastic material and soft silic...,[This Hybrid Protector Case is custom made for...,NaN,NextKin,[],"{""Manufacturer"": ""NextKin"", ""Is Discontinued B...",B00CDXINAQ,1
27,Computers,NVIDIA NVS 310 by PNY 1 GB DDR3 PCI Express Ge...,4.2,17,"[Chipset: NVIDIA NVS 310, Video Memory: 512MB ...",[This pre-owned or refurbished product has bee...,34.98,Amazon Renewed,"[Electronics, Computers & Accessories, Compute...","{""Product Dimensions"": ""5.7 x 2.5 x 2.8 inches...",B07QF5117Q,0


In [26]:
computers = meta[
    meta["main_category"] == "Computers"
].copy()

computers["categories_group"] = computers["categories"].apply(
    lambda x: " > ".join(map(str, x))
    if isinstance(x, (list, tuple, np.ndarray))
    else str(x)
)

In [29]:
computers_grouped = (
    computers
    .groupby("categories_group")["price"]
    .agg(
        total_products="count",
        unique_prices="nunique",
        min_price="min",
        median_price="median",
        max_price="max"
    )
    .reset_index()
)

display(computers_grouped)

,categories_group,total_products,unique_prices,min_price,median_price,max_price
0,,4436,1863,0.01,39.210,6999.00
1,Car & Vehicle Electronics > Vehicle Electronic...,0,0,NaN,NaN,NaN
2,Car & Vehicle Electronics > Vehicle Electronic...,0,0,NaN,NaN,NaN
3,Electronics > 12-Month Financing on Computers,1,1,219.99,219.990,219.99
4,Electronics > 2 in 1 PCs > Convertibles,0,0,NaN,NaN,NaN
...,...,...,...,...,...,...
763,Electronics > iPad Air 2 & iPad mini 3 Accesso...,0,0,NaN,NaN,NaN
764,Electronics > iPad Air 2 & iPad mini 3 Accesso...,1,1,24.99,24.990,24.99
765,Electronics > iPad Air Accessories,0,0,NaN,NaN,NaN
766,Electronics > iPad mini Accessories > Cases,0,0,NaN,NaN,NaN
